onnxruntime/core/providers/xnnpack/math/gemm.cc

```diff
@@ -0,0 +1,195 @@
+// Copyright(c) Microsoft Corporation.All rights reserved.
+// Licensed under the MIT License.
+
+#include "gemm.h"
+#include "core/graph/constants.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+#include "core/framework/transpose_helper.h"
+#include "core/providers/utils.h"
+#include "core/providers/xnnpack/detail/utils.h"
+#include "core/framework/tensorprotoutils.h"
+#include "gsl/gsl-lite.hpp"
+
+#include <iostream>
+
+    namespace onnxruntime {
+  namespace xnnpack {
+
+  bool Gemm::IsGemmOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+    bool supported = false;
+    const onnxruntime::Node& node = node_unit.GetNode();
+
+    // use do {} while(false) so it's easier to set a breakpoint on the return
+    do {
+      const auto& input_defs = node.InputDefs();
+
+      if (input_defs.size() < 2) {
+        break;
+      }
+
+      const auto& A_arg = *input_defs[0];
+      const auto& B_arg = *input_defs[1];
+      const auto& C_arg = *input_defs[2];
+
+      // we only support float currently
+      const auto* A_type = A_arg.TypeAsProto();
+      const auto* B_type = B_arg.TypeAsProto();
+      const auto* C_type = C_arg.TypeAsProto();
+
+      if (A_type == nullptr || B_type == nullptr || C_type == nullptr ||
+          A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ||
+          B_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ||
+          C_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+        break;
+      }
+
+      // B & C matrices must be constant
+      if (!graph.IsConstantInitializer(B_arg.Name(), true)) {
+        break;
+      }
+
+      if (!graph.IsConstantInitializer(C_arg.Name(), true)) {
+        break;
+      }
+
+      // making sure we are dealing with MatMul
+      const auto* A_shape = A_arg.Shape();
+      const auto* B_shape = B_arg.Shape();
+      const auto* C_shape = C_arg.Shape();
+      
+       if (!A_shape || A_shape->dim_size() > 3 || (A_shape->dim(0).dim_value() != 1 && A_shape->dim(1).dim_value() != 1)) {
+        break;
+      }
+
+      if (!B_shape || B_shape->dim_size() > 3 || (B_shape->dim(0).dim_value() != 1 && B_shape->dim(1).dim_value() != 1)) {
+        break;
+      }
+
+      if (!C_shape || C_shape->dim_size() > 3) {
+        break;
+      }
+
+      ProtoHelperNodeContext nc(node_unit.GetNode());
+      OpNodeProtoHelper info(&nc);
+
+      supported = true;
+
+    } while (false);
+
+    return supported;
+  }
+
+  Gemm::Gemm(const OpKernelInfo& info) : GemmBase(info), OpKernel(info) {
```

Comment:
Looks Gemm and Matmul have a lot similar code, Could we just imlemented them on the same class/file/baseclass, so we can leverage most of common code and save binary size in mobile-senario.

Reply:
Good idea - let me see how this can be implemented. there are a few challenges with this kind of integration but it make sense to combine the two. I will update once after the review